NetKet/Graph/pygraph.hpp

```diff
@@ -99,62 +138,55 @@ auto WithEdges(py::iterable xs, Function&& callback)
-    AbstractGraph::ColorMap colours;
-    if (hint != 0) {
-      colours.reserve(hint);
-    }
-    for (; first != last; ++first) {
-      auto const x = first->template cast<std::tuple<int, int, int>>();
-      auto const edge = make_edge(std::get<0>(x), std::get<1>(x));
-      edges.push_back(edge);
-      if (!colours.emplace(edge, std::get<2>(x)).second) {
-        // Failed to insert an edge because it already exists
-        throw InvalidInputError{"Edge list contains duplicates."};
-      }
-    }
+    auto colors = detail::Iterable2ColorMap(xs);
+    std::vector<AbstractGraph::Edge> edges;
+    edges.reserve(colors.size());
+    std::transform(
+        begin(colors), end(colors), std::back_inserter(edges),
+        [](std::pair<AbstractGraph::Edge, int> const& x) { return x.first; });
     return std::forward<Function>(callback)(std::move(edges),
-                                            std::move(colours));
+                                            std::move(colors));
+  } else {
+    return std::forward<Function>(callback)(detail::Iterable2Edges(xs));
   }
-  // No colours
-  for (; first != last; ++first) {
-    auto const x = first->template cast<std::tuple<int, int>>();
-    edges.push_back(make_edge(std::get<0>(x), std::get<1>(x)));
-  }
-  // NOTE(twesterhout): Yes, I know that this screws up the algorithmic
-  // complexity, but it's fast enough to be unnoticeable for all practical
-  // purposes.
-  std::sort(begin(edges), end(edges));
-  if (std::unique(begin(edges), end(edges)) != end(edges)) {
-    throw InvalidInputError{"Edge list contains duplicates."};
-  }
-  return std::forward<Function>(callback)(std::move(edges));
 }
+} // namespace
 
 void AddGraphModule(py::module& m) {
   auto subm = m.def_submodule("graph");
 
   py::class_<AbstractGraph>(subm, "Graph") ADDGRAPHMETHODS(AbstractGraph);
 
   py::class_<Hypercube, AbstractGraph>(subm, "Hypercube")
-
-      .def(py::init<int, int, bool, std::vector<std::vector<int>>>(),
-           py::arg("L"), py::arg("ndim"), py::arg("pbc") = true,
-           py::arg("edgecolors") = std::vector<std::vector<int>>())
-      .def("Nsites", &Hypercube::Nsites)
-      .def("AdjacencyList", &Hypercube::AdjacencyList)
-      .def("SymmetryTable", &Hypercube::SymmetryTable)
-      .def("EdgeColors", &Hypercube::EdgeColors)
-      .def("IsBipartite", &Hypercube::IsBipartite)
-      .def("IsConnected", &Hypercube::IsConnected)
-      .def("Distances", &Hypercube::Distances)
-      .def("AllDistances", &Hypercube::AllDistances) ADDGRAPHMETHODS(Hypercube);
+      .def(py::init<int, int, bool>(), py::arg("length"), py::arg("ndim") = 1,
+           py::arg("pbc") = true)
+      .def(py::init([](int const length, py::iterable xs) {
+             return make_unique<Hypercube>(length,
+                                           detail::Iterable2ColorMap(xs));
+           }),
+           py::arg("length"), py::arg("colors"))
+      .def_property_readonly("number_sites", &Hypercube::Nsites)
```

Comment:
Is it a good idea to have several ways to get the number of sites? Maybe we should decide between `len(graph)` and `graph.number_sites`.

Reply:
I provided both to show the available choices. Which one would you vote for? Personally, I like the `__len__`-based solution, but I have no rationale for it :)

Reply:
Just as with iteration, I think `len(graph)` cold be ambiguous (should it return the number of vertices or edges?). So personally I'd also prefer to be explicit here.

For comparison: [`networkx` implements both methods](https://networkx.github.io/documentation/latest/reference/classes/graph.html#counting-nodes-edges-and-neighbors). It has a function for the number of sites and implements `__len__` to mean the same thing.

(A comment on the name, to me `number_sites` sounds a bit strange . If we keep it, I'd either abbreviate it to `num_sites` or go the full way and say `.number_of_sites` which is similar to what `networkx` uses.)

Reply:
I don't like the `__len__` based solution very much, it's a bit ambiguous. We can keep it if you guys like it, I am not totally against it. 
What was wrong with the name `n_sites` anyway ? :)